File: src/staging/collision.rs
//! Detect overlaps between the merged `agentpack-bundle` and the user's `~/.claude/` extension
//! dirs. Claude loads both, which produces duplicate slash commands (e.g. `/code-tutor` and
//! `/agentpack-bundle:code-tutor`). By default we drop the staged pack copy and keep `~/.claude`.

use std::collections::HashSet;
use std::env;
use std::fs;
use std::io::{self, IsTerminal};
use std::path::Path;

use walkdir::WalkDir;

use crate::error::{AgentpackError, Result};
use crate::fs_util::remove_path_any;

/// Skip stripping pack copies when both exist (duplicated slash UX remains).
const IGNORE_ENV: &str = "AGENTPACK_IGNORE_USER_BUNDLE_COLLISION";

/// Lowercase skill slugs removed from staged harness trees so `verify_staging` can skip them.
pub(super) struct StagingCollisionRemoval {
    pub skill_slugs_lower: HashSet<String>,
}

fn eprint_collision_warning(line: &str) {
    if io::stderr().is_terminal() {
        eprintln!("\x1b[33m⚠ {line}\x1b[0m");
    } else {
        eprintln!("warning: {line}");
    }
}

fn collect_skill_slugs(skills_dir: &Path, out: &mut HashSet<String>) -> Result<()> {
    if !skills_dir.is_dir() {
        return Ok(());
    }
    for e in fs::read_dir(skills_dir).map_err(|e| AgentpackError::io(skills_dir, e))? {
        let e = e.map_err(|e| AgentpackError::io(skills_dir, e))?;
        let p = e.path();
        if p.is_dir() {
            if let Some(name) = p.file_name().and_then(|s| s.to_str()) {
                out.insert(name.to_lowercase());
            }
        }
    }
    Ok(())
}

fn collect_md_stems_recursive(dir: &Path, out: &mut HashSet<String>) -> Result<()> {
    if !dir.is_dir() {
        return Ok(());
    }
    for e in fs::read_dir(dir).map_err(|e| AgentpackError::io(dir, e))? {
        let e = e.map_err(|e| AgentpackError::io(dir, e))?;
        let p = e.path();
        if p.is_dir() {
            collect_md_stems_recursive(&p, out)?;
        } else if p
            .extension()
            .map(|ext| ext.eq_ignore_ascii_case("md"))
            .unwrap_or(false)
        {
            if let Some(stem) = p.file_stem().and_then(|s| s.to_str()) {
                out.insert(stem.to_lowercase());
            }
        }
    }
    Ok(())
}

fn remove_skill_slug_dir(skills_root: &Path, slug_lower: &str) -> Result<()> {
    if !skills_root.is_dir() {
        return Ok(());
    }
    for e in fs::read_dir(skills_root).map_err(|e| AgentpackError::io(skills_root, e))? {
        let e = e.map_err(|e| AgentpackError::io(skills_root, e))?;
        let p = e.path();
        if p.is_dir() {
            if let Some(name) = p.file_name().and_then(|s| s.to_str()) {
                if name.to_lowercase() == slug_lower {
                    remove_path_any(&p)?;
                }
            }
        }
    }
    Ok(())
}

fn remove_md_stems_under_tree(root: &Path, stem_lower: &str) -> Result<()> {
    if !root.is_dir() {
        return Ok(());
    }
    for entry in WalkDir::new(root).follow_links(false).contents_first(true) {
        let entry = entry.map_err(|e| AgentpackError::Staging(e.to_string()))?;
        let p = entry.path();
        if !p.is_file() {
            continue;
        }
        if !p
            .extension()
            .map(|ext| ext.eq_ignore_ascii_case("md"))
            .unwrap_or(false)
        {
            continue;
        }
        let lower = p
            .file_stem()
            .and_then(|s| s.to_str())
            .map(|s| s.to_lowercase());
        if lower.as_deref() != Some(stem_lower) {
            continue;
        }
        remove_path_any(p)?;
    }
    Ok(())
}

/// Like **[`resolve_user_claude_bundle_collisions`]** but uses **`home_dir`** as the user profile
/// root (directory that contains **`.claude`**). **`None`** skips resolution (same as missing real home).
pub(super) fn resolve_user_claude_bundle_collisions_with_home(
    bundle: &Path,
    opencode: &Path,
    codex: &Path,
    cursor_pack: &Path,
    home_dir: Option<&Path>,
) -> Result<StagingCollisionRemoval> {
    let no_removals = || StagingCollisionRemoval {
        skill_slugs_lower: HashSet::new(),
    };

    if env::var(IGNORE_ENV).ok().as_deref() == Some("1") {
        tracing::warn!("skipping bundle vs ~/.claude collision handling ({IGNORE_ENV}=1)");
        return Ok(no_removals());
    }

    let Some(home) = home_dir else {
        return Ok(no_removals());
    };
    let uc = home.join(".claude");

    let mut user_skills = HashSet::new();
    collect_skill_slugs(&uc.join("skills"), &mut user_skills)?;
    let mut user_cmd = HashSet::new();
    collect_md_stems_recursive(&uc.join("commands"), &mut user_cmd)?;
    let mut user_agents = HashSet::new();
    collect_md_stems_recursive(&uc.join("agents"), &mut user_agents)?;

    if user_skills.is_empty() && user_cmd.is_empty() && user_agents.is_empty() {
        return Ok(no_removals());
    }

    let mut removed_skills = HashSet::new();
    let harness_roots = [bundle, opencode, codex, cursor_pack];

    let mut bundle_skills = HashSet::new();
    collect_skill_slugs(&bundle.join("skills"), &mut bundle_skills)?;
    let mut bundle_cmd = HashSet::new();
    collect_md_stems_recursive(&bundle.join("commands"), &mut bundle_cmd)?;
    let mut bundle_agents = HashSet::new();
    collect_md_stems_recursive(&bundle.join("agents"), &mut bundle_agents)?;

    // Skills: remove matching slug dirs from all harness roots.
    let mut skill_keys: Vec<&String> = user_skills.intersection(&bundle_skills).collect();
    skill_keys.sort();
    for k in skill_keys {
        removed_skills.insert(k.clone());
        eprint_collision_warning(&format!(
            "Using ~/.claude skill `{k}`; omitted pack duplicate from staged bundle (and other harness trees)"
        ));
        for root in &harness_roots {
            remove_skill_slug_dir(&root.join("skills"), k)?;
        }
    }

    // Commands and agents: remove matching .md stems from applicable harness roots.
    // Codex doesn't have commands/ or agents/ trees, so we skip it.
    let md_roots = [bundle, opencode, cursor_pack];
    for (user_set, bundle_set, dir_name, label) in [
        (&user_cmd, &bundle_cmd, "commands", "command"),
        (&user_agents, &bundle_agents, "agents", "agent"),
    ] {
        let mut keys: Vec<&String> = user_set.intersection(bundle_set).collect();
        keys.sort();
        for k in keys {
            eprint_collision_warning(&format!(
                "Using ~/.claude {label} `{k}`; omitted pack duplicate from staged bundle (and other harness trees)"
            ));
            for root in &md_roots {
                remove_md_stems_under_tree(&root.join(dir_name), k)?;
            }
        }
    }

    Ok(StagingCollisionRemoval {
        skill_slugs_lower: removed_skills,
    })
}

/// Remove staged pack copies that duplicate `~/.claude` skills / commands / agents so the user
/// install wins. Prints yellow warnings to stderr. When **`IGNORE_ENV=1`**, does nothing (duplicates
/// remain). Returns lowercase skill slugs removed from **`skills/`** for staging verification.
pub(super) fn resolve_user_claude_bundle_collisions(
    bundle: &Path,
    opencode: &Path,
    codex: &Path,
    cursor_pack: &Path,
) -> Result<StagingCollisionRemoval> {
    resolve_user_claude_bundle_collisions_with_home(
        bundle,
        opencode,
        codex,
        cursor_pack,
        dirs::home_dir().as_deref(),
    )
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::tempdir;

    #[test]
    fn no_user_dirs_no_op() {
        let t = tempdir().unwrap();
        fs::create_dir_all(t.path().join(".claude/skills")).unwrap();
        let b = t.path().join("b");
        fs::create_dir_all(&b).unwrap();
        assert!(super::resolve_user_claude_bundle_collisions_with_home(
            &b,
            &b,
            &b,
            &b,
            Some(t.path()),
        )
        .unwrap()
        .skill_slugs_lower
        .is_empty());
    }

    #[test]
    fn collision_removes_pack_skill() {
        let t = tempdir().unwrap();
        let uc = t.path().join(".claude/skills/code-tutor");
        fs::create_dir_all(&uc).unwrap();
        fs::write(uc.join("SKILL.md"), "---\nname: x\n---\n").unwrap();
        let bundle = t.path().join("bundle");
        let bundle_skill = bundle.join("skills/code-tutor");
        fs::create_dir_all(&bundle_skill).unwrap();
        fs::write(bundle_skill.join("SKILL.md"), "---\n---\n").unwrap();
        let op = t.path().join("op");
        fs::create_dir_all(op.join("skills/code-tutor")).unwrap();
        fs::write(op.join("skills/code-tutor/SKILL.md"), "x").unwrap();

        let r = super::resolve_user_claude_bundle_collisions_with_home(
            &bundle,
            &op,
            &op,
            &op,
            Some(t.path()),
        )
        .unwrap();
        assert!(r.skill_slugs_lower.contains("code-tutor"));
        assert!(!bundle_skill.join("SKILL.md").exists());
        assert!(!op.join("skills/code-tutor/SKILL.md").exists());
        assert!(uc.join("SKILL.md").is_file());
    }

    #[test]
    fn collision_removes_pack_agent_md() {
        let t = tempdir().unwrap();
        let user_ag = t.path().join(".claude/agents/foo.md");
        fs::create_dir_all(user_ag.parent().unwrap()).unwrap();
        fs::write(&user_ag, "---\n---\n").unwrap();
        let bundle = t.path().join("bundle");
        fs::create_dir_all(bundle.join("agents")).unwrap();
        fs::write(bundle.join("agents/foo.md"), "---\n---\n").unwrap();

        super::resolve_user_claude_bundle_collisions_with_home(
            &bundle,
            &bundle,
            &bundle,
            &bundle,
            Some(t.path()),
        )
        .unwrap();
        assert!(!bundle.join("agents/foo.md").exists());
        assert!(user_ag.is_file());
    }
}

File: internal/staging/collision.go
package staging

import (
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

type CollisionRemoval struct{ SkillSlugs map[string]struct{} }

func ResolveCollisionsWithHome(bundle string, skillRoots, markdownRoots []string, home string) (CollisionRemoval, error) {
	removed := CollisionRemoval{SkillSlugs: make(map[string]struct{})}
	if home == "" {
		return removed, nil
	}
	userSkills, err := directoryNames(filepath.Join(home, ".claude", "skills"), filepath.Join(home, ".grok", "skills"))
	if err != nil {
		return removed, err
	}
	userCommands, err := markdownStems(filepath.Join(home, ".claude", "commands"), filepath.Join(home, ".grok", "commands"))
	if err != nil {
		return removed, err
	}
	userAgents, err := markdownStems(filepath.Join(home, ".claude", "agents"), filepath.Join(home, ".grok", "agents"))
	if err != nil {
		return removed, err
	}
	bundleSkills, err := directoryNames(filepath.Join(bundle, "skills"))
	if err != nil {
		return removed, err
	}
	bundleCommands, err := markdownStems(filepath.Join(bundle, "commands"))
	if err != nil {
		return removed, err
	}
	bundleAgents, err := markdownStems(filepath.Join(bundle, "agents"))
	if err != nil {
		return removed, err
	}
	for _, slug := range intersection(userSkills, bundleSkills) {
		removed.SkillSlugs[slug] = struct{}{}
		fmt.Fprintf(os.Stderr, "warning: Using user-installed skill `%s`; omitted pack duplicate from staged bundle (and other harness trees)\n", slug)
		for _, root := range skillRoots {
			if err := removeNamedDirectory(filepath.Join(root, "skills"), slug); err != nil {
				return removed, err
			}
		}
	}
	for _, collision := range []struct {
		user, pack       map[string]struct{}
		directory, label string
	}{{userCommands, bundleCommands, "commands", "command"}, {userAgents, bundleAgents, "agents", "agent"}} {
		for _, stem := range intersection(collision.user, collision.pack) {
			fmt.Fprintf(os.Stderr, "warning: Using user-installed %s `%s`; omitted pack duplicate from staged bundle (and other harness trees)\n", collision.label, stem)
			for _, root := range markdownRoots {
				if err := removeMarkdownStem(filepath.Join(root, collision.directory), stem); err != nil {
					return removed, err
				}
			}
		}
	}
	return removed, nil
}

func directoryNames(roots ...string) (map[string]struct{}, error) {
	result := make(map[string]struct{})
	for _, root := range roots {
		entries, err := os.ReadDir(root)
		if os.IsNotExist(err) {
			continue
		}
		if err != nil {
			return nil, err
		}
		for _, entry := range entries {
			if entry.IsDir() {
				result[strings.ToLower(entry.Name())] = struct{}{}
			}
		}
	}
	return result, nil
}

func markdownStems(roots ...string) (map[string]struct{}, error) {
	result := make(map[string]struct{})
	for _, root := range roots {
		if _, err := os.Stat(root); os.IsNotExist(err) {
			continue
		}
		err := filepath.WalkDir(root, func(path string, entry fs.DirEntry, err error) error {
			if err != nil {
				return err
			}
			if !entry.IsDir() && strings.EqualFold(filepath.Ext(entry.Name()), ".md") {
				result[strings.ToLower(strings.TrimSuffix(entry.Name(), filepath.Ext(entry.Name())))] = struct{}{}
			}
			return nil
		})
		if err != nil {
			return nil, err
		}
	}
	return result, nil
}

func intersection(left, right map[string]struct{}) []string {
	var values []string
	for value := range left {
		if _, exists := right[value]; exists {
			values = append(values, value)
		}
	}
	sort.Strings(values)
	return values
}

func removeNamedDirectory(root, lowerName string) error {
	entries, err := os.ReadDir(root)
	if os.IsNotExist(err) {
		return nil
	}
	if err != nil {
		return err
	}
	for _, entry := range entries {
		if entry.IsDir() && strings.EqualFold(entry.Name(), lowerName) {
			if err := os.RemoveAll(filepath.Join(root, entry.Name())); err != nil {
				return err
			}
		}
	}
	return nil
}

func removeMarkdownStem(root, lowerStem string) error {
	if _, err := os.Stat(root); os.IsNotExist(err) {
		return nil
	}
	return filepath.WalkDir(root, func(path string, entry fs.DirEntry, err error) error {
		if err != nil {
			return err
		}
		if !entry.IsDir() && strings.EqualFold(filepath.Ext(entry.Name()), ".md") && strings.EqualFold(strings.TrimSuffix(entry.Name(), filepath.Ext(entry.Name())), lowerStem) {
			return os.Remove(path)
		}
		return nil
	})
}
